// @ant-design/pro-skeleton@1.0.13/es/component/Result/index.js
import "antd/es/space/style";
import _Space from "antd/es/space";
import "antd/es/skeleton/style";
import _Skeleton from "antd/es/skeleton";
import "antd/es/card/style";
import _Card from "antd/es/card";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { PageHeaderSkeleton } from '../List';

var ResultPageSkeleton = function ResultPageSkeleton(_ref) {
  var _ref$active = _ref.active,
      active = _ref$active === void 0 ? true : _ref$active,
      pageHeader = _ref.pageHeader;
  return _jsxs("div", {
    style: {
      width: '100%'
    },
    children: [pageHeader !== false && _jsx(PageHeaderSkeleton, {
      active: active
    }), _jsx(_Card, {
      children: _jsxs("div", {
        style: {
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center',
          flexDirection: 'column',
          padding: 128
        },
        children: [_jsx(_Skeleton.Avatar, {
          size: 64,
          style: {
            marginBottom: 32
          }
        }), _jsx(_Skeleton.Button, {
          active: active,
          style: {
            width: 214,
            marginBottom: 8
          }
        }), _jsx(_Skeleton.Button, {
          active: active,
          style: {
            width: 328
          },
          size: "small"
        }), _jsxs(_Space, {
          style: {
            marginTop: 24
          },
          children: [_jsx(_Skeleton.Button, {
            active: active,
            style: {
              width: 116
            }
          }), _jsx(_Skeleton.Button, {
            active: active,
            style: {
              width: 116
            }
          })]
        })]
      })
    })]
  });
};

export default ResultPageSkeleton;
