// @antv/l7-component@2.22.3/lib/utils/anchor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.anchorType = exports.anchorTranslate = void 0;
exports.applyAnchorClass = applyAnchorClass;
let anchorType = exports.anchorType = /*#__PURE__*/function (anchorType) {
  anchorType["CENTER"] = "center";
  anchorType["TOP"] = "top";
  anchorType["TOP-LEFT"] = "top-left";
  anchorType["TOP-RIGHT"] = "top-right";
  anchorType["BOTTOM"] = "bottom";
  anchorType["BOTTOM-LEFT"] = "bottom-left";
  anchorType["LEFT"] = "left";
  anchorType["RIGHT"] = "right";
  return anchorType;
}({});
const anchorTranslate = exports.anchorTranslate = {
  center: 'translate(-50%,-50%)',
  top: 'translate(-50%,0)',
  'top-left': 'translate(0,0)',
  'top-right': 'translate(-100%,0)',
  bottom: 'translate(-50%,-100%)',
  'bottom-left': 'translate(0,-100%)',
  'bottom-right': 'translate(-100%,-100%)',
  left: 'translate(0,-50%)',
  right: 'translate(-100%,-50%)'
};
function applyAnchorClass(element, anchor, prefix) {
  const classList = element.classList;
  for (const key in anchorTranslate) {
    if (anchorTranslate.hasOwnProperty(key)) {
      classList.remove(`l7-${prefix}-anchor-${key}`);
    }
  }
  classList.add(`l7-${prefix}-anchor-${anchor}`);
}
