// @antv/l7-map@2.22.3/lib/map/handler/click_zoom.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ClickZoomHandler = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _transformProvider = require("./transform-provider");
/**
 * The `ClickZoomHandler` allows the user to zoom the map at a point by double clicking
 * It is used by other handlers
 */
class ClickZoomHandler {
  /** @internal */
  constructor(map) {
    (0, _defineProperty2.default)(this, "_tr", void 0);
    (0, _defineProperty2.default)(this, "_enabled", void 0);
    (0, _defineProperty2.default)(this, "_active", void 0);
    this._tr = new _transformProvider.TransformProvider(map);
    this.reset();
  }
  reset() {
    this._active = false;
  }
  dblclick(e, point) {
    e.preventDefault();
    return {
      cameraAnimation: map => {
        map.easeTo({
          duration: 300,
          zoom: this._tr.zoom + (e.shiftKey ? -1 : 1),
          around: this._tr.unproject(point)
        }, {
          originalEvent: e
        });
      }
    };
  }
  enable() {
    this._enabled = true;
  }
  disable() {
    this._enabled = false;
    this.reset();
  }
  isEnabled() {
    return this._enabled;
  }
  isActive() {
    return this._active;
  }
}
exports.ClickZoomHandler = ClickZoomHandler;
