// @antv/l7-map@2.22.3/lib/map/handler/one_finger_touch_drag.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateOneFingerTouchRotationHandler = exports.generateOneFingerTouchPitchHandler = void 0;
var _drag_handler = require("./drag_handler");
var _drag_move_state_manager = require("./drag_move_state_manager");
const assignEvents = handler => {
  handler.touchstart = handler.dragStart;
  handler.touchmoveWindow = handler.dragMove;
  handler.touchend = handler.dragEnd;
};
const generateOneFingerTouchRotationHandler = ({
  enable,
  clickTolerance,
  bearingDegreesPerPixelMoved = 0.8
}) => {
  const touchMoveStateManager = new _drag_move_state_manager.OneFingerTouchMoveStateManager();
  return new _drag_handler.DragHandler({
    clickTolerance,
    move: (lastPoint, point) => ({
      bearingDelta: (point.x - lastPoint.x) * bearingDegreesPerPixelMoved
    }),
    moveStateManager: touchMoveStateManager,
    enable,
    assignEvents
  });
};
exports.generateOneFingerTouchRotationHandler = generateOneFingerTouchRotationHandler;
const generateOneFingerTouchPitchHandler = ({
  enable,
  clickTolerance,
  pitchDegreesPerPixelMoved = -0.5
}) => {
  const touchMoveStateManager = new _drag_move_state_manager.OneFingerTouchMoveStateManager();
  return new _drag_handler.DragHandler({
    clickTolerance,
    move: (lastPoint, point) => ({
      pitchDelta: (point.y - lastPoint.y) * pitchDegreesPerPixelMoved
    }),
    moveStateManager: touchMoveStateManager,
    enable,
    assignEvents
  });
};
exports.generateOneFingerTouchPitchHandler = generateOneFingerTouchPitchHandler;
