// node-import-ts@1.0.8/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const ts = __importStar(require("typescript"));
const compile_1 = __importDefault(require("./compile"));
const path_1 = __importStar(require("path"));
const import_fresh_1 = __importDefault(require("import-fresh"));
module.exports = (filePath) => {
    const outDir = (0, path_1.join)(__dirname, '../.cache');
    const fileName = path_1.default.basename(filePath).replace('.ts', '.js');
    const index = (0, compile_1.default)([filePath], {
        noEmitOnError: false,
        noImplicitAny: false,
        allowJs: true,
        target: ts.ScriptTarget.ES5,
        module: ts.ModuleKind.CommonJS,
        outDir,
        skipLibCheck: true,
    });
    if (index !== 0) {
        throw new Error('编译失败');
    }
    const msg = (0, import_fresh_1.default)((0, path_1.join)(outDir, fileName));
    return msg.default ? msg.default : msg;
};
