// @antv/l7-layers@2.22.3/lib/utils/blend.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BlendTypes = void 0;
var _l7Core = require("@antv/l7-core");
const BlendTypes = exports.BlendTypes = {
  [_l7Core.BlendType.additive]: {
    enable: true,
    func: {
      srcRGB: _l7Core.gl.ONE,
      dstRGB: _l7Core.gl.ONE,
      srcAlpha: 1,
      dstAlpha: 1
    }
  },
  [_l7Core.BlendType.none]: {
    enable: false
  },
  [_l7Core.BlendType.normal]: {
    enable: true,
    func: {
      srcRGB: _l7Core.gl.SRC_ALPHA,
      dstRGB: _l7Core.gl.ONE_MINUS_SRC_ALPHA,
      srcAlpha: 1,
      dstAlpha: 1
    }
  },
  [_l7Core.BlendType.subtractive]: {
    enable: true,
    func: {
      srcRGB: _l7Core.gl.ONE,
      dstRGB: _l7Core.gl.ONE,
      srcAlpha: _l7Core.gl.ZERO,
      dstAlpha: _l7Core.gl.ONE_MINUS_SRC_COLOR
    },
    equation: {
      rgb: _l7Core.gl.FUNC_SUBTRACT,
      alpha: _l7Core.gl.FUNC_SUBTRACT
    }
  },
  [_l7Core.BlendType.max]: {
    enable: true,
    func: {
      srcRGB: _l7Core.gl.ONE,
      dstRGB: _l7Core.gl.ONE
    },
    equation: {
      rgb: _l7Core.gl.MAX_EXT
    }
  },
  [_l7Core.BlendType.min]: {
    enable: true,
    func: {
      srcRGB: _l7Core.gl.ONE,
      dstRGB: _l7Core.gl.ONE
    },
    equation: {
      rgb: _l7Core.gl.MIN_EXT
    }
  }
};
