// @antv/l7-component@2.22.3/lib/control/exportImage.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ExportImage = void 0;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _icon = require("../utils/icon");
var _buttonControl = _interopRequireDefault(require("./baseControl/buttonControl"));
class ExportImage extends _buttonControl.default {
  constructor(...args) {
    var _this;
    super(...args);
    _this = this;
    (0, _defineProperty2.default)(this, "onClick", /*#__PURE__*/(0, _asyncToGenerator2.default)(function* () {
      const {
        onExport
      } = _this.controlOption;
      onExport === null || onExport === void 0 || onExport(yield _this.getImage());
    }));
    /**
     * 将多张图片合并为一张图片
     * @protected
     * @param base64List
     */
    (0, _defineProperty2.default)(this, "mergeImage", /*#__PURE__*/(0, _asyncToGenerator2.default)(function* (...base64List) {
      var _this$mapsService$get, _this$mapsService$get2;
      const {
        imageType
      } = _this.controlOption;
      const {
        width = 0,
        height = 0
      } = (_this$mapsService$get = (_this$mapsService$get2 = _this.mapsService.getContainer()) === null || _this$mapsService$get2 === void 0 ? void 0 : _this$mapsService$get2.getBoundingClientRect()) !== null && _this$mapsService$get !== void 0 ? _this$mapsService$get : {};
      const canvas = document.createElement('canvas');
      canvas.width = width;
      canvas.height = height;
      const context = canvas.getContext('2d');
      const imgList = yield Promise.all(base64List.map(base64 => {
        return new Promise(resolve => {
          const img = new Image();
          img.onload = () => {
            resolve(img);
          };
          img.src = base64;
        });
      }));
      imgList.forEach(img => {
        context === null || context === void 0 || context.drawImage(img, 0, 0, width, height);
      });
      return canvas.toDataURL(`image/${imageType}`);
    }));
  }
  onAdd() {
    const button = super.onAdd();
    button.addEventListener('click', this.onClick);
    return button;
  }
  getDefault(option) {
    return (0, _objectSpread2.default)((0, _objectSpread2.default)({}, super.getDefault(option)), {}, {
      title: '导出图片',
      btnIcon: (0, _icon.createL7Icon)('l7-icon-export-picture'),
      imageType: 'png'
    });
  }
  getImage() {
    var _this2 = this;
    return (0, _asyncToGenerator2.default)(function* () {
      const mapImage = yield _this2.mapsService.exportMap('png');
      const layerImage = yield _this2.scene.exportPng('png');
      return _this2.mergeImage(
      // 在 Map 底图模式下 mapImage 为 undefined
      ...[mapImage, layerImage].filter(base64 => base64));
    })();
  }
}
exports.default = exports.ExportImage = ExportImage;
