// @antv/l7-component@2.22.3/lib/control/logo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Logo = void 0;
var _l7Utils = require("@antv/l7-utils");
var _baseControl = require("./baseControl");
class Logo extends _baseControl.Control {
  getDefault() {
    return {
      position: _baseControl.PositionType.BOTTOMLEFT,
      name: 'logo',
      href: 'https://l7.antv.antgroup.com/',
      img: 'https://gw.alipayobjects.com/mdn/rms_816329/afts/img/A*GRb1TKp4HcMAAAAAAAAAAAAAARQnAQ'
    };
  }
  onAdd() {
    const container = _l7Utils.DOM.create('div', 'l7-control-logo');
    this.setLogoContent(container);
    return container;
  }
  onRemove() {
    return null;
  }
  setOptions(option) {
    super.setOptions(option);
    if (this.checkUpdateOption(option, ['img', 'href'])) {
      _l7Utils.DOM.clearChildren(this.container);
      this.setLogoContent(this.container);
    }
  }
  setLogoContent(container) {
    const {
      href,
      img
    } = this.controlOption;
    const imgDOM = _l7Utils.DOM.create('img');
    imgDOM.setAttribute('src', img);
    imgDOM.setAttribute('aria-label', 'AntV logo');
    _l7Utils.DOM.setUnDraggable(imgDOM);
    if (href) {
      const anchorDOM = _l7Utils.DOM.create('a', 'l7-control-logo-link');
      anchorDOM.target = '_blank';
      anchorDOM.href = href;
      anchorDOM.rel = 'noopener nofollow';
      anchorDOM.setAttribute('rel', 'noopener nofollow');
      anchorDOM.appendChild(imgDOM);
      container.appendChild(anchorDOM);
    } else {
      container.appendChild(imgDOM);
    }
  }
}
exports.default = exports.Logo = Logo;
