// @antv/l7-component@2.22.3/lib/control/mouseLocation.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.MouseLocation = void 0;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _l7Core = require("@antv/l7-core");
var _l7Utils = require("@antv/l7-utils");
var _control = _interopRequireDefault(require("./baseControl/control"));
class MouseLocation extends _control.default {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "location", [0, 0]);
    (0, _defineProperty2.default)(this, "onMouseMove", e => {
      let position = this.location;
      // 适配不同底图，事件返回的数据名称不一致
      const lngLat = e.lngLat || e.lnglat;
      const {
        transform
      } = this.controlOption;
      if (lngLat) {
        position = [lngLat.lng, lngLat.lat];
      }
      this.location = position;
      if (transform) {
        position = transform(position);
      }
      this.insertLocation2HTML(position);
      this.emit('locationChange', position);
    });
  }
  getLocation() {
    return this.location;
  }
  getDefault(option) {
    return (0, _objectSpread2.default)((0, _objectSpread2.default)({}, super.getDefault(option)), {}, {
      position: _l7Core.PositionType.BOTTOMLEFT,
      transform: ([lng, lat]) => {
        return [+(+lng).toFixed(6), +(+lat).toFixed(6)];
      }
    });
  }
  onAdd() {
    const container = _l7Utils.DOM.create('div', 'l7-control-mouse-location');
    container.innerHTML = '&nbsp;';
    this.mapsService.on('mousemove', this.onMouseMove);
    return container;
  }
  onRemove() {
    this.mapsService.off('mousemove', this.onMouseMove);
  }
  insertLocation2HTML(position) {
    this.container.innerText = position.join(', ');
  }
}
exports.default = exports.MouseLocation = MouseLocation;
