// @antv/l7-source@2.22.3/lib/utils/tile/getRasterData.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/utils/tile/getRasterData.ts
var getRasterData_exports = {};
__export(getRasterData_exports, {
  getRasterFile: () => getRasterFile
});
module.exports = __toCommonJS(getRasterData_exports);
var import_l7_utils = require("@antv/l7-utils");
var import_bands = require("../bandOperation/bands");
var import_request = require("./request");
var getRasterFile = (tile, requestParameters, callback, rasterFormat, operation) => __async(void 0, null, function* () {
  const tileBandParams = (0, import_request.getTileBandParams)(requestParameters.url);
  if (tileBandParams.length > 1) {
    const { rasterFiles, xhrList, errList } = yield getMultiArrayBuffer(
      tileBandParams,
      requestParameters
    );
    (0, import_request.bindCancel)(tile, xhrList);
    if (errList.length > 0) {
      callback(errList, null);
      return;
    }
    (0, import_bands.processRasterData)(rasterFiles, rasterFormat, operation, callback);
  } else {
    const xhr = (0, import_l7_utils.getArrayBuffer)(requestParameters, (err, imgData) => {
      if (err) {
        callback(err);
      } else if (imgData) {
        const rasterFiles = [
          {
            data: imgData,
            bands: tileBandParams[0].bands
          }
        ];
        (0, import_bands.processRasterData)(rasterFiles, rasterFormat, operation, callback);
      }
    });
    (0, import_request.bindCancel)(tile, [xhr]);
  }
});
function getMultiArrayBuffer(tileBandParams, requestParameters) {
  return __async(this, null, function* () {
    const rasterFiles = [];
    const xhrList = [];
    const errList = [];
    for (let i = 0; i < tileBandParams.length; i++) {
      const tileBandParam = tileBandParams[i];
      const params = __spreadProps(__spreadValues({}, requestParameters), {
        url: tileBandParam.url
      });
      const bands = tileBandParam.bands;
      const { err, data, xhr } = yield (0, import_l7_utils.makeXMLHttpRequestPromise)(__spreadProps(__spreadValues({}, params), {
        type: "arrayBuffer"
      }));
      if (err) {
        errList.push(err);
      }
      xhrList.push(xhr);
      rasterFiles.push({
        data,
        bands
      });
    }
    return { rasterFiles, xhrList, errList };
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getRasterFile
});
