// @antv/l7-layers@2.22.3/lib/geometry/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _BaseLayer = _interopRequireDefault(require("../core/BaseLayer"));
var _models = _interopRequireDefault(require("./models"));
class GeometryLayer extends _BaseLayer.default {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "type", 'GeometryLayer');
    (0, _defineProperty2.default)(this, "defaultSourceConfig", {
      data: [{
        x: 0,
        y: 0
      }],
      options: {
        parser: {
          type: 'json',
          x: 'x',
          y: 'y'
        }
      }
    });
  }
  buildModels() {
    var _this = this;
    return (0, _asyncToGenerator2.default)(function* () {
      const modelType = _this.getModelType();
      _this.layerModel = new _models.default[modelType](_this);
      yield _this.initLayerModels();
    })();
  }
  getDefaultConfig() {
    const type = this.getModelType();
    const defaultConfig = {
      plane: {},
      sprite: {},
      billboard: {}
    };
    return defaultConfig[type];
  }
  getModelType() {
    var _shapeAttribute$scale;
    const shapeAttribute = this.styleAttributeService.getLayerStyleAttribute('shape');
    const shape = shapeAttribute === null || shapeAttribute === void 0 || (_shapeAttribute$scale = shapeAttribute.scale) === null || _shapeAttribute$scale === void 0 ? void 0 : _shapeAttribute$scale.field;
    if (shape === 'plane') {
      return 'plane';
    } else if (shape === 'sprite') {
      return 'sprite';
    } else if (shape === 'billboard') {
      return 'billboard';
    } else {
      return 'plane';
    }
  }
}
exports.default = GeometryLayer;
