// @antv/l7-core@2.22.3/lib/services/renderer/passes/MultiPassRenderer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _IMultiPassRenderer = require("../IMultiPassRenderer");
/**
 * ported from Three.js EffectComposer
 * @example
 * const renderer = new MultiPassRenderer([
 *   new ClearPass(),
 *   new RenderPass({
 *     models: [
 *       new Model(),
 *       new Model(),
 *     ],
 *   }),
 *   new CopyPass({
 *     renderToScreen: true,
 *   }),
 *   new TAAPass(),
 * ]);
 * renderer.render();
 * @see https://yuque.antfin-inc.com/yuqi.pyq/fgetpa/apuvbf#dRM8W
 */
class MultiPassRenderer {
  constructor(postProcessor) {
    (0, _defineProperty2.default)(this, "passes", []);
    (0, _defineProperty2.default)(this, "layer", void 0);
    (0, _defineProperty2.default)(this, "renderFlag", void 0);
    (0, _defineProperty2.default)(this, "width", 0);
    (0, _defineProperty2.default)(this, "height", 0);
    this.postProcessor = postProcessor;
  }
  setLayer(layer) {
    this.layer = layer;
  }
  setRenderFlag(renderFlag) {
    this.renderFlag = renderFlag;
  }
  getRenderFlag() {
    return this.renderFlag;
  }
  getPostProcessor() {
    return this.postProcessor;
  }
  render() {
    var _this = this;
    return (0, _asyncToGenerator2.default)(function* () {
      for (const pass of _this.passes) {
        yield pass.render(_this.layer);
      }
      // this.layer.renderModels();
      yield _this.postProcessor.render(_this.layer);
    })();
  }
  resize(width, height) {
    if (this.width !== width || this.height !== height) {
      this.postProcessor.resize(width, height);
      this.width = width;
      this.height = height;
    }
  }
  add(pass, config) {
    if (pass.getType() === _IMultiPassRenderer.PassType.PostProcessing) {
      this.postProcessor.add(pass, this.layer, config);
    } else {
      pass.init(this.layer, config);
      this.passes.push(pass);
    }
  }
  insert(pass, config, index) {
    pass.init(this.layer, config);
    this.passes.splice(index, 0, pass);
  }
  destroy() {
    this.passes.length = 0;
  }
}
exports.default = MultiPassRenderer;
