// @antv/l7-core@2.22.3/lib/services/renderer/passes/RenderPass.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _IMultiPassRenderer = require("../IMultiPassRenderer");
var _BaseNormalPass = _interopRequireDefault(require("./BaseNormalPass"));
/**
 * RenderPass，负责输出到后续 PostProcessor 的 readFBO 中
 */
class RenderPass extends _BaseNormalPass.default {
  getType() {
    return _IMultiPassRenderer.PassType.Normal;
  }
  getName() {
    return 'render';
  }
  init(layer, config) {
    super.init(layer, config);
  }
  render(layer) {
    const {
      useFramebuffer,
      clear
    } = this.rendererService;
    const readFBO = layer.multiPassRenderer.getPostProcessor().getReadFBO();
    useFramebuffer(readFBO, () => {
      clear({
        color: [0, 0, 0, 0],
        depth: 1,
        stencil: 0,
        framebuffer: readFBO
      });
      // render to post processor
      layer.multiPassRenderer.setRenderFlag(false);
      // layer.render();

      layer.models.forEach(model => {
        model.draw({
          uniforms: layer.layerModel.getUninforms()
        });
      });
      layer.multiPassRenderer.setRenderFlag(true);
    });
  }
}
exports.default = RenderPass;
