// @antv/l7-renderer@2.22.3/lib/device/constants.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/device/constants.ts
var constants_exports = {};
__export(constants_exports, {
  blendEquationMap: () => blendEquationMap,
  blendFuncMap: () => blendFuncMap,
  cullFaceMap: () => cullFaceMap,
  depthFuncMap: () => depthFuncMap,
  hintMap: () => hintMap,
  primitiveMap: () => primitiveMap,
  sizeFormatMap: () => sizeFormatMap,
  stencilFuncMap: () => stencilFuncMap,
  stencilOpMap: () => stencilOpMap,
  typedArrayCtorMap: () => typedArrayCtorMap,
  wrapModeMap: () => wrapModeMap
});
module.exports = __toCommonJS(constants_exports);
var import_g_device_api = require("@antv/g-device-api");
var import_l7_core = require("@antv/l7-core");
var typedArrayCtorMap = {
  [import_l7_core.gl.FLOAT]: Float32Array,
  [import_l7_core.gl.UNSIGNED_BYTE]: Uint8Array,
  [import_l7_core.gl.SHORT]: Int16Array,
  [import_l7_core.gl.UNSIGNED_SHORT]: Uint16Array,
  [import_l7_core.gl.INT]: Int32Array,
  [import_l7_core.gl.UNSIGNED_INT]: Uint32Array
};
var primitiveMap = {
  [import_l7_core.gl.POINTS]: import_g_device_api.PrimitiveTopology.POINTS,
  [import_l7_core.gl.LINES]: import_g_device_api.PrimitiveTopology.LINES,
  [import_l7_core.gl.LINE_LOOP]: import_g_device_api.PrimitiveTopology.LINES,
  [import_l7_core.gl.LINE_STRIP]: import_g_device_api.PrimitiveTopology.LINE_STRIP,
  [import_l7_core.gl.TRIANGLES]: import_g_device_api.PrimitiveTopology.TRIANGLES,
  [import_l7_core.gl.TRIANGLE_FAN]: import_g_device_api.PrimitiveTopology.TRIANGLES,
  [import_l7_core.gl.TRIANGLE_STRIP]: import_g_device_api.PrimitiveTopology.TRIANGLE_STRIP
};
var sizeFormatMap = {
  [1]: import_g_device_api.Format.F32_R,
  [2]: import_g_device_api.Format.F32_RG,
  [3]: import_g_device_api.Format.F32_RGB,
  [4]: import_g_device_api.Format.F32_RGBA
};
var hintMap = {
  [import_l7_core.gl.STATIC_DRAW]: import_g_device_api.BufferFrequencyHint.STATIC,
  [import_l7_core.gl.DYNAMIC_DRAW]: import_g_device_api.BufferFrequencyHint.DYNAMIC,
  [import_l7_core.gl.STREAM_DRAW]: import_g_device_api.BufferFrequencyHint.DYNAMIC
};
var wrapModeMap = {
  [import_l7_core.gl.REPEAT]: import_g_device_api.AddressMode.REPEAT,
  [import_l7_core.gl.CLAMP_TO_EDGE]: import_g_device_api.AddressMode.CLAMP_TO_EDGE,
  [import_l7_core.gl.MIRRORED_REPEAT]: import_g_device_api.AddressMode.MIRRORED_REPEAT
};
var depthFuncMap = {
  [import_l7_core.gl.NEVER]: import_g_device_api.CompareFunction.NEVER,
  [import_l7_core.gl.ALWAYS]: import_g_device_api.CompareFunction.ALWAYS,
  [import_l7_core.gl.LESS]: import_g_device_api.CompareFunction.LESS,
  [import_l7_core.gl.LEQUAL]: import_g_device_api.CompareFunction.LEQUAL,
  [import_l7_core.gl.GREATER]: import_g_device_api.CompareFunction.GREATER,
  [import_l7_core.gl.GEQUAL]: import_g_device_api.CompareFunction.GEQUAL,
  [import_l7_core.gl.EQUAL]: import_g_device_api.CompareFunction.EQUAL,
  [import_l7_core.gl.NOTEQUAL]: import_g_device_api.CompareFunction.NOTEQUAL
};
var cullFaceMap = {
  [import_l7_core.gl.FRONT]: import_g_device_api.CullMode.FRONT,
  [import_l7_core.gl.BACK]: import_g_device_api.CullMode.BACK
};
var blendEquationMap = {
  [import_l7_core.gl.FUNC_ADD]: import_g_device_api.BlendMode.ADD,
  [import_l7_core.gl.MIN_EXT]: import_g_device_api.BlendMode.MIN,
  [import_l7_core.gl.MAX_EXT]: import_g_device_api.BlendMode.MAX,
  [import_l7_core.gl.FUNC_SUBTRACT]: import_g_device_api.BlendMode.SUBSTRACT,
  [import_l7_core.gl.FUNC_REVERSE_SUBTRACT]: import_g_device_api.BlendMode.REVERSE_SUBSTRACT
};
var blendFuncMap = {
  [import_l7_core.gl.ZERO]: import_g_device_api.BlendFactor.ZERO,
  [import_l7_core.gl.ONE]: import_g_device_api.BlendFactor.ONE,
  [import_l7_core.gl.SRC_COLOR]: import_g_device_api.BlendFactor.SRC,
  [import_l7_core.gl.ONE_MINUS_SRC_COLOR]: import_g_device_api.BlendFactor.ONE_MINUS_SRC,
  [import_l7_core.gl.SRC_ALPHA]: import_g_device_api.BlendFactor.SRC_ALPHA,
  [import_l7_core.gl.ONE_MINUS_SRC_ALPHA]: import_g_device_api.BlendFactor.ONE_MINUS_SRC_ALPHA,
  [import_l7_core.gl.DST_COLOR]: import_g_device_api.BlendFactor.DST,
  [import_l7_core.gl.ONE_MINUS_DST_COLOR]: import_g_device_api.BlendFactor.ONE_MINUS_DST,
  [import_l7_core.gl.DST_ALPHA]: import_g_device_api.BlendFactor.DST_ALPHA,
  [import_l7_core.gl.ONE_MINUS_DST_ALPHA]: import_g_device_api.BlendFactor.ONE_MINUS_DST_ALPHA,
  [import_l7_core.gl.CONSTANT_COLOR]: import_g_device_api.BlendFactor.CONST,
  [import_l7_core.gl.ONE_MINUS_CONSTANT_COLOR]: import_g_device_api.BlendFactor.ONE_MINUS_CONSTANT,
  [import_l7_core.gl.CONSTANT_ALPHA]: import_g_device_api.BlendFactor.CONST,
  [import_l7_core.gl.ONE_MINUS_CONSTANT_ALPHA]: import_g_device_api.BlendFactor.ONE_MINUS_CONSTANT,
  [import_l7_core.gl.SRC_ALPHA_SATURATE]: import_g_device_api.BlendFactor.SRC_ALPHA_SATURATE
};
var stencilOpMap = {
  [import_l7_core.gl.REPLACE]: import_g_device_api.StencilOp.REPLACE,
  [import_l7_core.gl.KEEP]: import_g_device_api.StencilOp.KEEP,
  [import_l7_core.gl.ZERO]: import_g_device_api.StencilOp.ZERO,
  [import_l7_core.gl.INVERT]: import_g_device_api.StencilOp.INVERT,
  [import_l7_core.gl.INCR]: import_g_device_api.StencilOp.INCREMENT_CLAMP,
  [import_l7_core.gl.DECR]: import_g_device_api.StencilOp.DECREMENT_CLAMP,
  [import_l7_core.gl.INCR_WRAP]: import_g_device_api.StencilOp.INCREMENT_WRAP,
  [import_l7_core.gl.DECR_WRAP]: import_g_device_api.StencilOp.DECREMENT_WRAP
};
var stencilFuncMap = {
  [import_l7_core.gl.ALWAYS]: import_g_device_api.CompareFunction.ALWAYS,
  [import_l7_core.gl.EQUAL]: import_g_device_api.CompareFunction.EQUAL,
  [import_l7_core.gl.GEQUAL]: import_g_device_api.CompareFunction.GEQUAL,
  [import_l7_core.gl.GREATER]: import_g_device_api.CompareFunction.GREATER,
  [import_l7_core.gl.LEQUAL]: import_g_device_api.CompareFunction.LEQUAL,
  [import_l7_core.gl.LESS]: import_g_device_api.CompareFunction.LESS,
  [import_l7_core.gl.NEVER]: import_g_device_api.CompareFunction.NEVER,
  [import_l7_core.gl.NOTEQUAL]: import_g_device_api.CompareFunction.NOTEQUAL
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  blendEquationMap,
  blendFuncMap,
  cullFaceMap,
  depthFuncMap,
  hintMap,
  primitiveMap,
  sizeFormatMap,
  stencilFuncMap,
  stencilOpMap,
  typedArrayCtorMap,
  wrapModeMap
});
