// @antv/l7-renderer@2.22.3/lib/device/DeviceElements.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/device/DeviceElements.ts
var DeviceElements_exports = {};
__export(DeviceElements_exports, {
  default: () => DeviceElements
});
module.exports = __toCommonJS(DeviceElements_exports);
var import_g_device_api = require("@antv/g-device-api");
var import_l7_core = require("@antv/l7-core");
var import_constants = require("./constants");
var import_typedarray = require("./utils/typedarray");
var DeviceElements = class {
  constructor(device, options) {
    const { data, type, count = 0 } = options;
    let typed;
    if ((0, import_typedarray.isTypedArray)(data)) {
      typed = data;
    } else {
      typed = new import_constants.typedArrayCtorMap[this.type || import_l7_core.gl.UNSIGNED_INT](data);
    }
    this.type = type;
    this.count = count;
    this.indexBuffer = device.createBuffer({
      viewOrSize: typed,
      usage: import_g_device_api.BufferUsage.INDEX
    });
  }
  get() {
    return this.indexBuffer;
  }
  subData({
    data
  }) {
    let typed;
    if ((0, import_typedarray.isTypedArray)(data)) {
      typed = data;
    } else {
      typed = new import_constants.typedArrayCtorMap[this.type || import_l7_core.gl.UNSIGNED_INT](data);
    }
    this.indexBuffer.setSubData(0, new Uint8Array(typed.buffer));
  }
  destroy() {
    this.indexBuffer.destroy();
  }
};
