// @antv/l7-renderer@2.22.3/lib/device/DeviceTexture2D.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/device/DeviceTexture2D.ts
var DeviceTexture2D_exports = {};
__export(DeviceTexture2D_exports, {
  default: () => DeviceTexture2D,
  isTexture2D: () => isTexture2D
});
module.exports = __toCommonJS(DeviceTexture2D_exports);
var import_g_device_api = require("@antv/g-device-api");
var import_l7_core = require("@antv/l7-core");
var import_constants = require("./constants");
var import_typedarray = require("./utils/typedarray");
function isTexture2D(t) {
  return !!(t && t["texture"]);
}
var DeviceTexture2D = class {
  constructor(device, options) {
    this.device = device;
    this.options = options;
    this.isDestroy = false;
    const {
      wrapS = import_l7_core.gl.CLAMP_TO_EDGE,
      wrapT = import_l7_core.gl.CLAMP_TO_EDGE,
      aniso,
      mag = import_l7_core.gl.NEAREST,
      min = import_l7_core.gl.NEAREST
    } = options;
    this.createTexture(options);
    this.sampler = device.createSampler({
      addressModeU: import_constants.wrapModeMap[wrapS],
      addressModeV: import_constants.wrapModeMap[wrapT],
      minFilter: min === import_l7_core.gl.NEAREST ? import_g_device_api.FilterMode.POINT : import_g_device_api.FilterMode.BILINEAR,
      magFilter: mag === import_l7_core.gl.NEAREST ? import_g_device_api.FilterMode.POINT : import_g_device_api.FilterMode.BILINEAR,
      mipmapFilter: import_g_device_api.MipmapFilterMode.NO_MIP,
      // lodMinClamp: 0,
      // lodMaxClamp: 0,
      maxAnisotropy: aniso
    });
  }
  createTexture(options) {
    const {
      type = import_l7_core.gl.UNSIGNED_BYTE,
      width,
      height,
      flipY = false,
      format = import_l7_core.gl.RGBA,
      alignment = 1,
      usage = import_l7_core.TextureUsage.SAMPLED,
      // premultiplyAlpha = false,
      unorm = false,
      // colorSpace = gl.BROWSER_DEFAULT_WEBGL,
      // x = 0,
      // y = 0,
      // copy = false,
      label
    } = options;
    let { data } = options;
    this.width = width;
    this.height = height;
    let pixelFormat = import_g_device_api.Format.U8_RGBA_RT;
    if (type === import_l7_core.gl.UNSIGNED_BYTE && format === import_l7_core.gl.RGBA) {
      pixelFormat = unorm ? import_g_device_api.Format.U8_RGBA_NORM : import_g_device_api.Format.U8_RGBA_RT;
    } else if (type === import_l7_core.gl.UNSIGNED_BYTE && format === import_l7_core.gl.LUMINANCE) {
      pixelFormat = import_g_device_api.Format.U8_LUMINANCE;
    } else if (type === import_l7_core.gl.FLOAT && format === import_l7_core.gl.LUMINANCE) {
      pixelFormat = import_g_device_api.Format.F32_LUMINANCE;
    } else if (type === import_l7_core.gl.FLOAT && format === import_l7_core.gl.RGB) {
      if (this.device.queryVendorInfo().platformString === "WebGPU") {
        if (data) {
          data = (0, import_typedarray.extend3ChannelsTo4)(data, 0);
        }
        pixelFormat = import_g_device_api.Format.F32_RGBA;
      } else {
        pixelFormat = import_g_device_api.Format.F32_RGB;
      }
    } else if (type === import_l7_core.gl.FLOAT && format === import_l7_core.gl.RGBA) {
      pixelFormat = import_g_device_api.Format.F32_RGBA;
    } else if (type === import_l7_core.gl.FLOAT && format === import_l7_core.gl.RED) {
      pixelFormat = import_g_device_api.Format.F32_R;
    } else {
      throw new Error(`create texture error, type: ${type}, format: ${format}`);
    }
    this.texture = this.device.createTexture({
      format: pixelFormat,
      width,
      height,
      usage: usage === import_l7_core.TextureUsage.SAMPLED ? import_g_device_api.TextureUsage.SAMPLED : import_g_device_api.TextureUsage.RENDER_TARGET,
      pixelStore: {
        unpackFlipY: flipY,
        packAlignment: alignment
      },
      // mipLevelCount: usage === TextureUsage.RENDER_TARGET ? 1 : mipmap ? 1 : 0,
      mipLevelCount: 1
    });
    if (label) {
      this.device.setResourceName(this.texture, label);
    }
    if (data) {
      this.texture.setImageData([data]);
    }
  }
  get() {
    return this.texture;
  }
  update(props) {
    const { data } = props;
    this.texture.setImageData([data]);
  }
  bind() {
  }
  resize({ width, height }) {
    if (this.width !== width || this.height !== height) {
      this.destroy();
    }
    this.options.width = width;
    this.options.height = height;
    this.createTexture(this.options);
    this.isDestroy = false;
  }
  getSize() {
    return [this.width, this.height];
  }
  destroy() {
    var _a;
    if (!this.isDestroy && !this.texture.destroyed) {
      (_a = this.texture) == null ? void 0 : _a.destroy();
    }
    this.isDestroy = true;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  isTexture2D
});
