// @antv/l7-layers@2.22.3/lib/wind/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _BaseLayer = _interopRequireDefault(require("../core/BaseLayer"));
var _models = _interopRequireDefault(require("./models"));
class WindLayer extends _BaseLayer.default {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "type", 'WindLayer');
  }
  buildModels() {
    var _this = this;
    return (0, _asyncToGenerator2.default)(function* () {
      const modelType = _this.getModelType();
      _this.layerModel = new _models.default[modelType](_this);
      yield _this.initLayerModels();
    })();
  }
  renderModels(options = {}) {
    if (this.layerModel) {
      this.layerModel.render(options); // 独立的渲染流程
    }
    return this;
  }
  getDefaultConfig() {
    const type = this.getModelType();
    const defaultConfig = {
      wind: {}
    };
    return defaultConfig[type];
  }
  getModelType() {
    return 'wind';
  }
}
exports.default = WindLayer;
