// @antv/l7-layers@2.22.3/lib/tile/tile/DebugTile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _line = _interopRequireDefault(require("../../line"));
var _point = _interopRequireDefault(require("../../point"));
var _Tile = _interopRequireDefault(require("./Tile"));
class DebugTile extends _Tile.default {
  initTileLayer() {
    var _this = this;
    return (0, _asyncToGenerator2.default)(function* () {
      const sourceOptions = _this.getSourceOption();
      const pointData = sourceOptions.data.features[0].properties;
      const lineLayer = new _line.default().source(sourceOptions.data, sourceOptions.options).size(1).shape('line').color('red');
      const pointLayer = new _point.default({
        minZoom: _this.z - 1,
        maxZoom: _this.z + 1,
        textAllowOverlap: true
      }).source([pointData], {
        parser: {
          type: 'json',
          x: 'x',
          y: 'y'
        }
      }).size(20).color('red').shape(_this.key).style({
        stroke: '#fff',
        strokeWidth: 2
      });
      yield _this.addLayer(lineLayer);
      yield _this.addLayer(pointLayer);
      _this.isLoaded = true;
    })();
  }
  getSourceOption() {
    const rawSource = this.parent.getSource();
    const features = this.sourceTile.data.layers.testTile.features;
    return {
      data: {
        type: 'FeatureCollection',
        features
      },
      options: {
        parser: {
          type: 'geojson'
        },
        transforms: rawSource.transforms
      }
    };
  }
}
exports.default = DebugTile;
