// @antv/l7-layers@2.22.3/lib/tile/tile/RasterRGBTile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _raster = _interopRequireDefault(require("../../raster"));
var _Tile = _interopRequireDefault(require("./Tile"));
const _excluded = ["rasterData"];
class RasterTile extends _Tile.default {
  initTileLayer() {
    var _this = this;
    return (0, _asyncToGenerator2.default)(function* () {
      const attributes = _this.parent.getLayerAttributeConfig();
      const layerOptions = _this.getLayerOptions();
      const sourceOptions = _this.getSourceOption();
      const layer = new _raster.default((0, _objectSpread2.default)({}, layerOptions)).source(sourceOptions.data, sourceOptions.options);

      // 初始化数据映射
      // tslint:disable-next-line: no-unused-expression
      attributes && Object.keys(attributes).forEach(type => {
        var _attributes$attr, _attributes$attr2;
        const attr = type;
        // @ts-ignore
        layer[attr]((_attributes$attr = attributes[attr]) === null || _attributes$attr === void 0 ? void 0 : _attributes$attr.field, (_attributes$attr2 = attributes[attr]) === null || _attributes$attr2 === void 0 ? void 0 : _attributes$attr2.values);
      });
      yield _this.addLayer(layer);
      _this.isLoaded = true;
    })();
  }
  getSourceOption() {
    const rawSource = this.parent.getSource();
    const _this$sourceTile$data = this.sourceTile.data.data,
      {
        rasterData
      } = _this$sourceTile$data,
      res = (0, _objectWithoutProperties2.default)(_this$sourceTile$data, _excluded);
    return {
      data: rasterData,
      // 栅格数
      options: {
        parser: (0, _objectSpread2.default)({
          type: 'rasterRgb',
          extent: this.sourceTile.bounds
        }, res),
        transforms: rawSource.transforms
      }
    };
  }
}
exports.default = RasterTile;
