// @ant-design/pro-skeleton@1.0.13/lib/component/Result/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd/lib/space/style");

var _space = _interopRequireDefault(require("antd/lib/space"));

require("antd/lib/skeleton/style");

var _skeleton = _interopRequireDefault(require("antd/lib/skeleton"));

require("antd/lib/card/style");

var _card = _interopRequireDefault(require("antd/lib/card"));

var _jsxRuntime = require("react/jsx-runtime");

var _List = require("../List");

var ResultPageSkeleton = function ResultPageSkeleton(_ref) {
  var _ref$active = _ref.active,
      active = _ref$active === void 0 ? true : _ref$active,
      pageHeader = _ref.pageHeader;
  return (0, _jsxRuntime.jsxs)("div", {
    style: {
      width: '100%'
    },
    children: [pageHeader !== false && (0, _jsxRuntime.jsx)(_List.PageHeaderSkeleton, {
      active: active
    }), (0, _jsxRuntime.jsx)(_card.default, {
      children: (0, _jsxRuntime.jsxs)("div", {
        style: {
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center',
          flexDirection: 'column',
          padding: 128
        },
        children: [(0, _jsxRuntime.jsx)(_skeleton.default.Avatar, {
          size: 64,
          style: {
            marginBottom: 32
          }
        }), (0, _jsxRuntime.jsx)(_skeleton.default.Button, {
          active: active,
          style: {
            width: 214,
            marginBottom: 8
          }
        }), (0, _jsxRuntime.jsx)(_skeleton.default.Button, {
          active: active,
          style: {
            width: 328
          },
          size: "small"
        }), (0, _jsxRuntime.jsxs)(_space.default, {
          style: {
            marginTop: 24
          },
          children: [(0, _jsxRuntime.jsx)(_skeleton.default.Button, {
            active: active,
            style: {
              width: 116
            }
          }), (0, _jsxRuntime.jsx)(_skeleton.default.Button, {
            active: active,
            style: {
              width: 116
            }
          })]
        })]
      })
    })]
  });
};

var _default = ResultPageSkeleton;
exports.default = _default;
